### ISSUE TITLE
walkdir is type unstable

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#26030]
```jl
julia> function f()
             for (root, dirs, files) in walkdir(".")
             println("Directories in $root")
             for dir in dirs
                 println(joinpath(root, dir)) # path to directories
             end
             println("Files in $root")
             for file in files
                 println(joinpath(root, file)) # path to files
             end
         end
       end
f (generic function with 1 method)

julia> @code_warntype f()
Variables:
  #temp#@_2::Base.ChannelIterState{Any}
  root::Any
  dirs::Any
  files::Any
  #temp#@_6::Any
  #temp#@_7::Any
  #temp#@_8::Any
  file::Any
  dir::Any
  v<optimized out>
```

Not cirtical but had some benchmarks that gave bad results because it didn't realize that `file` was a `String`.

**Comment by inkydragon** [comment]
Issue resolved on 1.6.3.

```
julia> @code_warntype f()
Variables
  #self#::Core.Const(f)
  @_2::Union{Nothing, Tuple{Tuple{String, Vector{String}, Vector{String}}, Nothing}}
  @_3::Union{Nothing, Tuple{String, Int64}}
  @_4::Union{Nothing, Tuple{String, Int64}}
  @_5::Int64
  files::Vector{String}
  dirs::Vector{String}
  root::String
  dir::String
  file::String

...
```

**Comment by KristofferC** [comment]
Great!
